File: Hardwaver/lib/services/memory/memory_manager.dart
import 'package:sqlite3/sqlite3.dart';
import 'package:path/path.dart' as p;
import 'package:path_provider/path_provider.dart';
import 'dart:ffi';
import 'package:sqlite3/open.dart';

import '../../config/app_config.dart';
import '../../models/message.dart';
import 'short_term.dart';
import 'mid_term.dart';
import 'long_term.dart';

class MemoryManager {
  final ShortTermMemory _shortTerm;
  final MidTermMemory _midTerm;
  final LongTermMemory _longTerm;
  final Database _db;

  MemoryManager._({
    required ShortTermMemory shortTerm,
    required MidTermMemory midTerm,
    required LongTermMemory longTerm,
    required Database db,
  })  : _shortTerm = shortTerm,
        _midTerm = midTerm,
        _longTerm = longTerm,
        _db = db;

  static Future<MemoryManager> create({
    int shortTermMaxSize = AppConfig.shortTermMaxMessages,
  }) async {
    open.overrideFor(OperatingSystem.windows, () => DynamicLibrary.open('sqlite3.dll'));

    final dir = await getApplicationDocumentsDirectory();
    final dbPath = p.join(dir.path, 'dimotalk_memory.db');
    final db = sqlite3.open(dbPath);

    db.execute('PRAGMA journal_mode=WAL;');
    db.execute('PRAGMA foreign_keys=ON;');

    MidTermMemory.init(db);
    LongTermMemory.init(db);

    final shortTerm = ShortTermMemory(maxSize: shortTermMaxSize);
    final midTerm = MidTermMemory(db);
    final longTerm = LongTermMemory(db);

    longTerm.forgetExpired();

    return MemoryManager._(
      shortTerm: shortTerm,
      midTerm: midTerm,
      longTerm: longTerm,
      db: db,
    );
  }

  ShortTermMemory get shortTerm => _shortTerm;
  MidTermMemory get midTerm => _midTerm;
  LongTermMemory get longTerm => _longTerm;

  void addToShortTerm(Message msg) => _shortTerm.add(msg);

  Future<void> onSessionEnd({
    required String conversationId,
    required String userId,
    required String summary,
  }) async {
    await _midTerm.storeSummary(
      conversationId: conversationId,
      userId: userId,
      summary: summary,
    );
    _shortTerm.clear();
  }

  Future<void> dispose() async {
    _db.dispose();
  }
}

File: DimoTalk.Maui/Services/Memory/MemoryManager.cs
using DimoTalk.Maui.Config;
using DimoTalk.Maui.Models;
using Microsoft.Data.Sqlite;

namespace DimoTalk.Maui.Services.Memory;

public class MemoryManager : IDisposable
{
    public ShortTermMemory ShortTerm { get; }
    public MidTermMemory MidTerm { get; }
    public LongTermMemory LongTerm { get; }
    private readonly SqliteConnection _conn;

    private MemoryManager(ShortTermMemory shortTerm, MidTermMemory midTerm, LongTermMemory longTerm, SqliteConnection conn)
    {
        ShortTerm = shortTerm;
        MidTerm = midTerm;
        LongTerm = longTerm;
        _conn = conn;
    }

    public static async Task<MemoryManager> CreateAsync(string? userId = null, int shortTermMaxSize = AppConfig.ShortTermMaxMessages)
    {
        SQLitePCL.Batteries_V2.Init();

        var dir = FileSystem.AppDataDirectory;
        Directory.CreateDirectory(dir);
        var dbPath = Path.Combine(dir, "dimotalk_memory.db");

        var conn = new SqliteConnection($"Data Source={dbPath}");
        await conn.OpenAsync();

        using (var pragma = conn.CreateCommand())
        {
            pragma.CommandText = "PRAGMA journal_mode=WAL; PRAGMA foreign_keys=ON;";
            pragma.ExecuteNonQuery();
        }

        var shortTerm = new ShortTermMemory(shortTermMaxSize);
        var midTerm = new MidTermMemory(conn);
        var longTerm = new LongTermMemory(conn);
        longTerm.ForgetExpired();

        return new MemoryManager(shortTerm, midTerm, longTerm, conn);
    }

    public void AddToShortTerm(Message msg) => ShortTerm.Add(msg);

    public async Task OnSessionEndAsync(string conversationId, string userId, string summary)
    {
        MidTerm.StoreSummary(conversationId, userId, summary);
        ShortTerm.Clear();
        await Task.CompletedTask;
    }

    public void Dispose()
    {
        _conn.Dispose();
    }
}
